### ISSUE TITLE
docker run: docker client stuck in FIN_WAIT2

### ISSUE THREAD

**Issue by stelcheck** [docker/docker#25966]
**Output of `docker version`:**

```
Client:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:23:11 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:23:11 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 12
 Running: 10
 Paused: 0
 Stopped: 2
Images: 136
Server Version: 1.11.2
Storage Driver: devicemapper
 Pool Name: docker-253:0-768275-pool
 Pool Blocksize: 65.54 kB
 Base Device Size: 107.4 GB
 Backing Filesystem: xfs
 Data file: /dev/loop0
 Metadata file: /dev/loop1
 Data Space Used: 4.682 GB
 Data Space Total: 107.4 GB
 Data Space Available: 102.7 GB
 Metadata Space Used: 9.626 MB
 Metadata Space Total: 2.147 GB
 Metadata Space Available: 2.138 GB
 Udev Sync Supported: true
 Deferred Removal Enabled: false
 Deferred Deletion Enabled: false
 Deferred Deleted Device Count: 0
 Data loop file: /var/lib/docker/devicemapper/devicemapper/data
 WARNING: Usage of loopback devices is strongly discouraged for production use. Either use `--storage-opt dm.thinpooldev` or use `--storage-opt dm.no_warn_on_loop_devices=true` to suppress this warning.
 Metadata loop file: /var/lib/docker/devicemapper/devicemapper/metadata
 Library Version: 1.02.107-RHEL7 (2015-10-14)
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay bridge null host
Kernel Version: 3.10.0-327.4.4.el7.x86_64
Operating System: CentOS Linux 7 (Core)
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.481 GiB
Name: node1.docker1.x.x
ID: x
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Registry: https://index.docker.io/v1/
WARNING: bridge-nf-call-ip6tables is disabled
Cluster store: zk://x.x.x.x:2181,x.x.x.x:2181/swarm
Cluster advertise: x.x.x.x:2375
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Running from different client versions of Docker (no errors)

**Steps to reproduce the issue:**
1. Run `docker run` -- run a process that takes a long time to complete (1 minute or more)

**Describe the results you received:**

Process sometimes stall

**Describe the results you expected:**

Process shouldn't stall

**Additional information you deem important (e.g. issue happens only occasionally):**

The task does complete successfully, and the container in the swarm cluster does shutdown. The problem is that `docker run` doesn't actually quit.

It looks like the docker client CLI essentially gets stuck into an infinite loop:

```
client ~ # ps aauwx | grep docker
root     18639  0.0  0.0 103244   888 pts/0    S+   04:36   0:00 grep docker
user  18784  0.0  0.0 280540  7828 ?        Sl   Aug22   0:01 docker run ...

client ~ # strace -p18784
Process 18784 attached - interrupt to quit
futex(0x127d238, FUTEX_WAIT, 0, NULL
# CTRL-C

client ~ # lsof | grep 18784 | tail -n5

docker    18784 user    1w     FIFO                0,8       0t0  152669047 pipe
docker    18784 user    2w     FIFO                0,8       0t0  152669047 pipe
docker    18784 user    3u     IPv4          152669386       0t0        TCP clientip:39923->swarm:2376 (ESTABLISHED)
docker    18784 user    4u      REG                0,9         0       3827 [eventpoll]
docker    18784 user    5u     IPv4          152669406       0t0        TCP clientip:39932->swarm:2376 (FIN_WAIT2)

client ~ # netstat -apn | grep FIN
tcp        0      0 clientip:39932          swarm:2376           FIN_WAIT2   18784/docker
```

In other words, it looks like every once in a while, the client side doesn't receive notification that the server has closed it's socket. I checked on the swarm, and the connection has indeed been cleaned up.

We do not see any similar behaviour with any other services we are running on the same network.

**Comment by stelcheck** [comment]
I just did some additional research, so here is a quick update:

1. FIN_WAIT_2 might not be the issue after all. It appears that `docker run` does exit even if there is a FIN_WAIT_2 connection as listed above
2. However, it does look like the sockets are kept alive on the client side...
3. While I cannot see any matching connections open on the server side (netstat doesn't list it on the server).
4. The connection in FIN_WAIT_2 state continues to generate traffic (sent every 30 seconds):

```bash
06:56:05.202189 IP clientip.56290 > host: Flags [.], ack 1, win 4114, length 0
06:56:05.202219 IP swarm.2376 > clientip.56290: Flags [.], ack 1, win 235, options [nop,nop,TS val 679871121 ecr 2938290499], length 0
PPP
```

On this run, the client was on a macOS machine. According to dtruss, it would do select calls every minutes or so:

```bash
select(0x0, 0x0, 0x0, 0x0, 0x700000080DE8)		 = 0 0
```

**Comment by thaJeztah** [comment]
Sorry for the lack of response here @stelcheck. I'm on sure I fully grasp the issue, but ping @cpuguy83 perhaps you have some thoughts 😇

**Comment by cpuguy83** [comment]
This sounds like a bug in swarm's TCP connection hijack. Might be good to report over there in github.com/docker/swarm.
